dart/common/Uri.cpp

```diff
@@ -0,0 +1,435 @@
+        else
+          mQuery = _base.mQuery;
+      }
+      else
+      {
+        if(_relative.mPath->front() == '/')
+          mPath = removeDotSegments(*_relative.mPath);
+        else
+          mPath = removeDotSegments(mergePaths(_base, _relative));
+
+        mQuery = _relative.mQuery;
+      }
+
+      mAuthority = _base.mAuthority;
+    }
+
+    mScheme = _base.mScheme;
+  }
+
+  mFragment = _relative.mFragment;
+  return true;
+}
+
+bool Uri::fromString(const std::string& _input)
+{
+  // This is regex is from Appendix B of RFC 3986.
+  static regex uriRegex(
+    R"END(^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)END",
+    regex::extended | regex::optimize);
+  static const size_t schemeIndex = 2;
+  static const size_t authorityIndex = 4;
+  static const size_t pathIndex = 5;
+  static const size_t queryIndex = 7;
+  static const size_t fragmentIndex = 9;
+
+  clear();
+
+  smatch matches;
+  if(!regex_match(_input, matches, uriRegex))
+    return false;
+
+  assert(matches.size() > schemeIndex);
+  const ssub_match& schemeMatch = matches[schemeIndex];
+  if(schemeMatch.matched)
+    mScheme = schemeMatch;
+
+  assert(matches.size() > authorityIndex);
+  const ssub_match& authorityMatch = matches[authorityIndex];
+  if(authorityMatch.matched)
+    mAuthority = authorityMatch;
+
+  assert(matches.size() > pathIndex);
+  const ssub_match& pathMatch = matches[pathIndex];
+  if(pathMatch.matched)
+    mPath = pathMatch;
+
+  assert(matches.size() > queryIndex);
+  const ssub_match& queryMatch = matches[queryIndex];
+  if(queryMatch.matched)
+    mQuery = queryMatch;
+
+  assert(matches.size() > fragmentIndex);
+  const ssub_match& fragmentMatch = matches[fragmentIndex];
+  if (fragmentMatch.matched)
+    mFragment = fragmentMatch;
+
+  return true;
+}
+
+std::string Uri::toString() const
+{
+  // This function implements the pseudo-code from Section 5.3 of RFC 3986.
+  std::stringstream output;
+
+  if(mScheme)
+    output << *mScheme << ":";
+
+  if(mAuthority)
+    output << "//" << *mAuthority;
+
+  output << mPath.get_value_or("");
+
+  if(mQuery)
+    output << "?" << *mQuery;
+
+  if(mFragment)
+    output << "#" << *mFragment;
+
+  return output.str();
+}
+
+bool Uri::fromStringOrPath(const std::string& _input)
+{
+  if (!fromString(_input))
+    return false;
+
+  // Assume that any URI without a scheme is a path.
+  if (!mScheme && mPath)
+  {
+    mScheme = "file";
+
+    // Replace backslashes (from Windows paths) with forward slashes.
+    std::replace(std::begin(*mPath), std::end(*mPath), '\\', '/');
+  }
+
+  return true;
+}
+
+std::string Uri::getUri(const std::string& _input)
+{
+  Uri uri;
+  if(uri.fromStringOrPath(_input))
+    return uri.toString();
+  else
+    return "";
+}
+
+std::string Uri::getRelativeUri(
+  const std::string& _base, const std::string& _relative, bool _strict)
+{
+  Uri baseUri;
+  if(!baseUri.fromString(_base))
+  {
+    dtwarn << "[getRelativeUri] Failed parsing base URI '"
```

Comment:
Since this file contains two classes (``Uri`` and ``UriComponent``) it would probably be good to be explicit about the function ``[Uri::getRelativeUri]`` so the user knows more exactly where the warning is coming from.